vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/AbstractNumberField.java

```diff
@@ -57,14 +58,57 @@ public abstract class AbstractNumberField<C extends AbstractNumberField<C, T>, T
     private boolean stepSetByUser;
     private boolean minSetByUser;
 
-    private boolean manualValidationEnabled = false;
-
     private DomListenerRegistration inputListenerRegistration;
 
     private final CopyOnWriteArrayList<ValidationStatusChangeListener<T>> validationStatusChangeListeners = new CopyOnWriteArrayList<>();
 
-    private String customErrorMessage;
-    private String constraintErrorMessage;
+    private Validator<T> defaultValidator = (value, context) -> {
+        boolean hasBadInput = valueEquals(value, getEmptyValue())
+                && isInputValuePresent();
+        if (hasBadInput) {
+            return ValidationResult.error(getI18nErrorMessage(
+                    AbstractNumberFieldI18n::getBadInputErrorMessage));
+        }
+
+        if (context == null) {
```

Comment:
Unrelated, but this is puzzling without context (🥁). Might be worth a comment when context can be null, and why to exclude the required validator then.

Reply:
Good point. I've added a comment.